### ISSUE TITLE
Sample App using COM-Activated Toast Notifications + Squirrel

### ISSUE THREAD

**Issue by sopelt** [Squirrel/Squirrel.Windows#1435]
Hi, I am trying to integrate Windows 10 style notifications with COM activation (for being able to react to buttons/text/...) in an app already distributed via squirrel. I am trying to stay along the lines of https://docs.microsoft.com/en-us/windows/uwp/design/shell/tiles-and-notifications/send-local-toast-desktop and https://github.com/WindowsNotifications/desktop-toasts ... the sample uses WiX and that works but in our app the COM activation does not seem to fire. I am setting a consistent app user model ID (as discussed in https://github.com/Squirrel/Squirrel.Windows/issues/1338 ) at runtime and for toast handling as well as having the activator com class id as generated by squirrel (is there a way to inspect this metadata on the generated shortcut?) ... Does anyone happen to have a reference/sample for that scenario? A C#/.net app (not electron) that is squirrel packaged and uses Win10 notifications with interaction? It seems to be a somewhat unusual combination ;)

I am thinking of forking https://github.com/WindowsNotifications/desktop-toasts and adding a squirrel package to it to see if it keeps working or not ...

I am thankful for any input.

**Comment by egm0121** [comment]
@sopelt have you eventually found a solution? I'm hitting the same issue with an electron app that triggers notifications with buttons.

From my tests the core issue is that when the app gets installed via Squirrel, the Start Menu Shortcut that is required with a proper appUserModelID in order for notifications to work is actually pointing to the Squirrel exe, that takes care of launching the correct version of the app.

if you manually create a shortcut to your app exe directly it correctly works and does trigger the onActivate event.

Squirrel Start menu shortcut Target:
`C:\Users\User\AppData\Local\myapp\myapp.exe` - actions not activated

Manual Start menu shortcut Target:
`C:\Users\User\AppData\Local\myapp\app-6.8.0\myapp.exe` - actions work correctly

**Comment by sopelt** [comment]
@egm0121  .. no, unfortunately not. We kept the legacy balloons but they now seem broken (always getting closed event even when clicked) .. so I am also wondering on how to handle this.

If there is no way to activate without having a link on the actual binary the stub binary would have to handle that which I don't see happening :-/

**Comment by anaisbetts** [comment]
@sopelt Call https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-setcurrentprocessexplicitappusermodelid and set it to the `com.squirrel.NUGET_ID.EXENAME` name

**Comment by sopelt** [comment]
Hi @anaisbetts ! Thanks for chiming in. Are you saying that having the AUMID set on the running process will cause the com activation to work? What happens when the process is no longer running? Should spawning via the stub exe work as well?

Since my last post I debugged some more and [NotifyIcon.BalloonTipClicked](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.notifyicon.balloontipclicked) stops working as soon as I set the AUMID on the process :D 

@egm0121 which scenarios have you tried yet?

**Comment by anaisbetts** [comment]
Correct. I have no idea why you have to do this, and what You Have To Do to make this work keeps changing between Windows versions. Just call the method on startup

**Comment by egm0121** [comment]
@sopelt when testing this I made sure that the electron process does have the correct `App User Model Id` set via (https://www.electronjs.org/docs/api/app#appsetappusermodelidid-windows).

If the `appUserModelId` is not specified, the windows 10 Toast notifications with buttons ( ToastGeneric template) are not displayed at all.

**Comment by sopelt** [comment]
I had another look @egm0121 and it looks like the secret sauce I have been missing is that in addition to the AUMID the shortcut also contains the CLSID for the COM activator. I dug a bit and it looks like [this](https://github.com/Squirrel/Squirrel.Windows/blob/develop/src/Squirrel/Utility.cs#L707) implements [RFC 4122](https://tools.ietf.org/html/rfc4122#section-4.3) for name/hash-based Guids to derive the CLSID from the AUMID. Is that assumption correct @anaisbetts ? That means this is stable and can be relied upon?

I tried putting this all together in a reasonable example [here](https://github.com/software-architects/desktop-toasts) (and it only took a year of distance and a bit of grey hair). Please have a look/give it a try.
